### PR TITLE
mds: invalid id for client eviction is to be treated as success

### PR DESCRIPTION
**Pull request by vshankar**
Introduced-by: 0ef5941a2e79
Fixes: http://tracker.ceph.com/issues/68132

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by neesingh-rh — COMMENTED** [review]
I understand its breaking manila but don't you think client evict with invalid id needs to be handled, it seems like the command with invalid id is correct. Some user might again encounter this as a issue and we might have a ticket raised for this in future.

**Comment by vshankar** [comment]
> I understand its breaking manila but don't you think client evict with invalid id needs to be handled, it seems like the command with invalid id is correct. Some user might again encounter this as a issue and we might have a ticket raised for this in future.

Are you referring to the issue about all clients getting evicted or the specific issue of handling invalid client-id? Commit https://github.com/ceph/ceph/commit/0ef5941a2e79feeae79a705080b57d410fa38136 correctly fixes the client eviction part (where all clients were getting evicted), but the commit additionally changes the behaviour of returning -ESRCH on invalid client-id which breaks manila. Also, its good to have commands idempotent, so running the same command twice, which results in the same output is actually a good thing.

**Comment by neesingh-rh** [comment]
> > I understand its breaking manila but don't you think client evict with invalid id needs to be handled, it seems like the command with invalid id is correct. Some user might again encounter this as a issue and we might have a ticket raised for this in future.
> 
> Are you referring to the issue about all clients getting evicted or the specific issue of handling invalid client-id? Commit [0ef5941](https://github.com/ceph/ceph/commit/0ef5941a2e79feeae79a705080b57d410fa38136) correctly fixes the client eviction part (where all clients were getting evicted), but the commit additionally changes the behaviour of returning -ESRCH on invalid client-id which breaks manila. Also, its good to have commands idempotent, so running the same command twice, which results in the same output is actually a good thing.

The tracker issue was about eviction of all clients when wrong flags like `-h`  were passed which got fixed. I am referring to the issue of handling invalid client id. With this patch whether the `id` passed is valid or invalid , user doesn't get to know what happened...until and unless he checks the `client ls` output whether number of clients remain same or not, because we don't even get any succesfull message after client eviction. If checking `client ls` again and again after running command to determine the validness of `id` is fine, then this is good to go.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
> > > I understand its breaking manila but don't you think client evict with invalid id needs to be handled, it seems like the command with invalid id is correct. Some user might again encounter this as a issue and we might have a ticket raised for this in future.
> > 
> > 
> > Are you referring to the issue about all clients getting evicted or the specific issue of handling invalid client-id? Commit [0ef5941](https://github.com/ceph/ceph/commit/0ef5941a2e79feeae79a705080b57d410fa38136) correctly fixes the client eviction part (where all clients were getting evicted), but the commit additionally changes the behaviour of returning -ESRCH on invalid client-id which breaks manila. Also, its good to have commands idempotent, so running the same command twice, which results in the same output is actually a good thing.
> 
> The tracker issue was about eviction of all clients when wrong flags like `-h` were passed which got fixed. I am referring to the issue of handling invalid client id, which i tried to fix too. With this patch whether the `id` passed is valid or invalid , user doesn't get to know what happened...until and unless he checks the `client ls` output whether number of clients remain same or not, because we don't even get any succesfull message after client eviction. If checking `client ls` again and again after running command to determine the validness of `id` is fine, then this is good to go.

Manila has also designed their workflow based on `session evict` not returning an error for invalid id, so, we'll have to keep that bit as it is :/

I know its a bit awkward, but we need to be good citizens with our users and also my earlier point about idempotent operations.

**Review by neesingh-rh — APPROVED** [review]

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -3149,7 +3149,7 @@ void MDSRankDispatcher::evict_clients(
   dout(20) << __func__ << " matched " << victims.size() << " sessions" << dendl;
 
   if (victims.empty()) {
-    on_finish(-ESRCH, "no hosts match", outbl);
+    on_finish(0, {}, outbl);
```
Comment (batrick) [inline src/mds/MDSRank.cc:3152]: Any reason to not also say "no hosts match"?

Comment (vshankar) [reply src/mds/MDSRank.cc:3152]: That's acceptable too.

Comment (vshankar) [reply src/mds/MDSRank.cc]: Fixed.

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68179](https://tracker.ceph.com/issues/68179).

**Review by vshankar — COMMENTED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240923114534-debug

**Comment by vshankar** [comment]
jenkins test windows

**Pull request merged**